### PR TITLE
feat: pairwise MCP config translation between coding assistants

### LINKED ISSUE #79: feat: pairwise MCP config translation between coding assistants
## Problem

Developers who use multiple coding assistants — or who switch between them — must manually recreate their MCP server configurations for each assistant. The underlying MCP server definitions are structurally ~80% identical across assistants (server name, command, args, environment variables), but each assistant stores them in a different file, with different JSON schema, and with different key names.

### Example: The Same MCP Server in Four Formats

**GitHub Copilot** (`.vscode/mcp.json`):
`json
{
  "servers": {
    "github": {
      "command": "npx",
      "args": ["-y", "@modelcontextprotocol/server-github"],
      "env": { "GITHUB_PERSONAL_ACCESS_TOKEN": "ghp_xxx" }
    }
  }
}
`

**Claude Code** (`.claude/settings.local.json`):
`json
{
  "mcpServers": {
    "github": {
      "command": "npx",
      "args": ["-y", "@modelcontextprotocol/server-github"],
      "env": { "GITHUB_PERSONAL_ACCESS_TOKEN": "ghp_xxx" }
    }
  }
}
`

**OpenCode** (`opencode.json`):
`json
{
  "mcp": {
    "github": {
      "type": "local",
      "command": ["npx", "-y", "@modelcontextprotocol/server-github"],
      "environment": { "GITHUB_PERSONAL_ACCESS_TOKEN": "ghp_xxx" },
      "enabled": true
    }
  }
}
`

**Roo Code** (`.roo/mcp.json`):
`json
{
  "mcpServers": {
    "github": {
      "command": "npx",
      "args": ["-y", "@modelcontextprotocol/server-github"],
      "env": { "GITHUB_PERSONAL_ACCESS_TOKEN": "ghp_xxx" }
    }
  }
}
`

This is tedious, error-prone, and blocks adoption of multi-assistant workflows.

### User Scenarios

**Scenario 1 — Individual developer switching assistants:** I've been using Claude Code with 5 MCP servers configured. I'm trying OpenCode. I want to bring my MCP config across without manually rewriting it.

**Scenario 2 — Individual developer using multiple assistants:** I use Copilot in VS Code and Claude Code in the terminal. When I add an MCP server to one, I want it available in the other.

**Scenario 3 — Team lead standardising tooling:** My team uses a mix of assistants. I want to define MCP servers once and push the config to all supported assistants in our repo.

**Scenario 4 — Migration from Roo Code:** I've been using Roo Code with MCP servers from the Roo Code Marketplace. I'm moving to Copilot and want to bring my MCP configs.

### CLI Shape: Exploration

The CLI shape for this feature needs to be explored. Below are three options with trade-offs. The chosen approach should be validated during implementation.

#### Option A: Dedicated `mcp` Subcommand

`ash
# Translate all MCP servers from Claude to Copilot
code-minions mcp translate --from claude --to copilot

# Translate a specific server
code-minions mcp translate --from claude --to copilot --server github

# Translate to multiple targets
code-minions mcp translate --from claude --to copilot,opencode

# Dry run
code-minions mcp translate --from claude --to copilot --dry-run

# Force overwrite conflicts
code-minions mcp translate --from claude --to copilot --force

# List MCP servers detected for an assistant
code-minions mcp list --for claude

# Target a specific directory
code-minions mcp translate --from claude --to copilot --target ./my-project
`

**Pros:** Clear, purpose-built, discoverable. The `mcp` subcommand can grow to include other MCP operations later (e.g., `mcp install` from the canonical format in #78).
**Cons:** New top-level subcommand adds surface area to the CLI.

#### Option B: Extend the `transfer` Command (#63)

`ash
# Transfer skills/agents AND MCP configs
code-minions transfer --from claude --to copilot

# Transfer MCP configs only
code-minions transfer --from claude --to copilot --mcp-only

# Transfer skills/agents only (skip MCP)
code-minions transfer --from claude --to copilot --skip-mcp
`

**Pros:** Single command for all cross-assistant migration. Aligns with #63's scope.
**Cons:** Overloads `transfer` with two different operations (file copy vs JSON merge). MCP translation has different merge semantics — flags could get confusing.

#### Option C: Hybrid — `mcp` Subcommand, `transfer` Calls It

`ash
# Standalone MCP translation
code-minions mcp translate --from claude --to copilot

# Transfer includes MCP by default
code-minions transfer --from claude --to copilot
# (internally calls mcp translate as part of the transfer)
`

**Pros:** Best of both worlds. Dedicated command for MCP-only workflows; integrated into transfer for full migrations.
**Cons:** Two code paths that need to stay in sync.

### Recommendation

**Option C (Hybrid)** is recommended. It gives MCP translation a clear home (`mcp` subcommand) while ensuring the `transfer` command (#63) provides a complete migration experience. The `mcp` package (#78) provides shared translation logic used by both commands.

### Open Questions for CLI Shape

- Should `mcp list` show the parsed/normalised config, or the raw file contents?
- Should `mcp translate` support `--all-targets` to write to every supported assistant at once?
- Should there be an `mcp diff` command that shows what differs between two assistants' configs?
- How does `--json` output mode work for MCP operations? (Consistent with existing `--json` flag on other commands.)

### Field Mapping

| Canonical Field | Copilot (`.vscode/mcp.json`) | Claude Code | OpenCode (`opencode.json`) | Roo Code (`.roo/mcp.json`) |
|----------------|-------------------------------|-------------|------------------------------|------------------------------|
| Server name | Key in `servers` | Key in `mcpServers` | Key in `mcp` | Key in `mcpServers` |
| `command` | `command` (string) | `command` (string) | `command[0]` (first element of array) | `command` (string) |
| `args` | `args` (array) | `args` (array) | `command[1:]` (rest of array) | `args` (array) |
| `env` | `env` (object) | `env` (object) | `environment` (object) | `env` (object) |
| `url` (remote) | `url` | Not standard yet | `url` (with `type: remote`) | `url` |
| `headers` | `headers` | N/A | `headers` | `headers` |
| Enabled/disabled | Per-server toggle | Plugin settings | `enabled` (bool) | Per-server toggle |

### Translation Steps

1. **Read** source assistant's MCP config file
2. **Parse** into a normalised intermediate representation (same struct used by the canonical format in #78)
3. **Filter** by `--server` flag if specified
4. **Translate** to target assistant's schema
5. **Merge** into target assistant's existing config file (same merge logic as #78)
6. **Write** the updated config file
7. **Report** results (added, skipped, conflicts, warnings)

### Edge Cases

| Scenario | Behaviour |
|----------|-----------|
| Source config file doesn't exist | Error: `No MCP config found for claude at .claude/settings.local.json` |
| Source config exists but has no MCP servers | Warning: `No MCP servers found in Claude Code config` |
| Target config file doesn't exist | Create it with just the translated MCP servers |
| Target config has other non-MCP settings | Preserve them (read-modify-write) |
| Server exists in target with identical config | Skip: `Server 'github' already configured for copilot (identical)` |
| Server exists in target with different config | Skip by default, overwrite with `--force`: `Server 'github' already configured for copilot (conflict, use --force to overwrite)` |
| Remote MCP server → assistant that doesn't support remote | Warning: `Server 'remote-api' uses HTTP transport, which is not supported by [target]. Skipping.` |
| OpenCode `command` array → Copilot `command` + `args` | Split: `command[0]` → `command`, `command[1:]` → `args` |
| Copilot `command` + `args` → OpenCode `command` array | Join: `[command] + args` → `command` array |
| `env` contains secrets | Warning: `⚠️ MCP config may contain secrets. Review before committing.` |
| Source assistant not recognised | Error: `Unknown assistant 'foo'. Supported: copilot, claude, opencode, roo` |
| Source and target are the same | Error: `Source and target assistant cannot be the same` |

### Shared Infrastructure with #78

This feature shares significant infrastructure with the canonical MCP format feature (#78):

| Component | #78 (Canonical) | This Issue (Pairwise) | Shared? |
|-----------|-----------------|----------------------|---------|
| Normalised MCP struct | `mcp.yaml` → struct | Source config → struct | ✅ Same struct |
| Per-assistant translators | Struct → target JSON | Struct → target JSON | ✅ Same translators |
| Merge logic | Merge into target config | Merge into target config | ✅ Same merger |
| Source parsers | YAML parser (`mcp.yaml`) | JSON parsers (per-assistant) | ❌ Different — pairwise needs per-assistant readers |
| CLI command | Part of `install` | New `mcp` subcommand | ❌ Different entry points |

### New/Modified Files

| File | Change |
|------|--------|
| `internal/mcp/reader.go` (new) | Per-assistant MCP config readers (parse Copilot, Claude, OpenCode, Roo Code formats into normalised struct) |
| `internal/mcp/reader_test.go` (new) | Tests for each assistant's config parsing |
| `internal/cmd/mcp.go` (new) | `mcp` subcommand group |
| `internal/cmd/mcp_translate.go` (new) | `mcp translate` command implementation |
| `internal/cmd/mcp_list.go` (new) | `mcp list` command implementation |
| `internal/cmd/mcp_translate_test.go` (new) | Integration tests for translate flow |
| `internal/cmd/root.go` | Wire `mcp` subcommand |
| `internal/cmd/transfer.go` (future, #63) | Call MCP translation as part of full transfer |
| `internal/assistant/assistant.go` | Add `MCPConfigPath` and `MCPConfigKey` to `Config` (shared with #78) |

### Supported Translation Paths

All permutations of the four supported assistants:

| From \ To | Copilot | Claude | OpenCode | Roo Code |
|-----------|---------|--------|----------|----------|
| **Copilot** | — | ✅ | ✅ | ✅ |
| **Claude** | ✅ | — | ✅ | ✅ |
| **OpenCode** | ✅ | ✅ | — | ✅ |
| **Roo Code** | ✅ | ✅ | ✅ | — |

12 translation paths, but only 4 readers + 4 writers needed (not 12 pairwise converters) because of the normalised intermediate representation.

## Alternatives Considered

- **Only support canonical format (#78), not pairwise** — Forces users to author `mcp.yaml` first, then install. Doesn't help users who already have configs in one assistant and want them in another.
- **Shell script/one-liner** — `jq` can do simple JSON transforms, but doesn't handle merge semantics, OpenCode's array-based command format, or validation. Not user-friendly.
- **Separate `mcp-migrate` tool** — Could exist independently of code-minions. But MCP config management is closely related to code-minions' mission of cross-assistant portability, and shares infrastructure with #78.
- **Config symlinking** — Symlink one assistant's config to another. Doesn't work because the formats differ. Would only work if all assistants used identical schemas.

## Additional Context

- Research document: `docs/research/053-openclaw-integration.md` (Part 2: MCP Configuration Formats, Part 6: Cross-Provider MCP Portability)
- Depends on / shares infrastructure with: #78 (canonical MCP format — normalised struct, translators, merger)
- Related: #63 (transfer command — Option C integrates MCP translation into transfer)
- [Model Context Protocol specification](https://modelcontextprotocol.io/)
- [OpenCode MCP documentation](https://opencode.ai/docs/mcp-servers)
- [Roo Code MCP configuration](https://docs.roocode.com/features/marketplace)

## Acceptance Criteria

- [ ] CLI shape is decided (Option A, B, or C) and documented
- [ ] `code-minions mcp translate --from <assistant> --to <assistant>` reads source MCP config and writes to target format
- [ ] All 12 translation paths work (4 assistants × 3 targets each)
- [ ] Per-assistant readers correctly parse: `.vscode/mcp.json`, `.claude/settings.local.json`, `opencode.json`, `.roo/mcp.json`
- [ ] Translation handles field mapping differences (OpenCode `command` array vs `command` + `args`, `env` vs `environment`)
- [ ] Merge semantics: new servers added, identical servers skipped, conflicts respected with `--force`
- [ ] `--server` flag translates a specific server only
- [ ] `--dry-run` previews changes without writing
- [ ] `--target` specifies working directory (defaults to `.`)
- [ ] `--json` output mode supported
- [ ] `code-minions mcp list --for <assistant>` lists detected MCP servers
- [ ] Warnings for: empty env vars, potential secrets, unsupported transport types
- [ ] Errors for: missing source config, unknown assistant, same source/target
- [ ] Non-MCP content in config files is preserved
- [ ] Unit tests cover: all 4 readers, all 4 writers, field mapping edge cases, merge scenarios
- [ ] Integration tests cover: end-to-end translate for at least 2 permutations

### PR DESCRIPTION
**Pull request by willvelida**
## Summary

Add mcp translate and mcp list CLI commands that read MCP server configurations from one coding assistant's format and write them into another's. Supports **Copilot**, **Claude Code**, and **OpenCode**.

### `mcp translate`
Translates MCP server configs between assistants:
`
code-minions mcp translate --from copilot --to claude
code-minions mcp translate --from copilot --to claude,opencode
code-minions mcp translate --from copilot --to claude --server github
code-minions mcp translate --from copilot --to claude --dry-run
code-minions mcp translate --from copilot --to claude --force
`

### `mcp list`
Lists MCP servers configured for an assistant:
`
code-minions mcp list --for copilot
code-minions mcp list --for claude --json
`

## Flags
| Flag | Description |
|------|-------------|
| `--from` | Source assistant |
| `--to` | Target assistant(s), comma-separated |
| `--for` | Assistant to inspect (list only) |
| `--server` | Translate only a specific server |
| `--force` | Overwrite conflicting servers |
| `--dry-run` | Preview without writing |
| `--target` | Working directory |
| `--json` / `--verbose` / `--quiet` | Output modes |

## New Files
| File | Purpose |
|------|---------|
| `internal/mcp/reader.go` | Reader interface + CopilotReader, ClaudeReader, OpenCodeReader |
| `internal/mcp/reader_test.go` | Reader tests + cross-assistant round-trip tests |
| `internal/mcp/translate.go` | Translation engine (read → parse → filter → translate → merge → write) |
| `internal/mcp/translate_test.go` | End-to-end translation tests |
| `internal/cmd/mcp.go` | Parent mcp subcommand group |
| `internal/cmd/mcp_translate.go` | mcp translate command |
| `internal/cmd/mcp_list.go` | mcp list command |
| `internal/cmd/mcp_test.go` | CLI-level tests |

## Modified Files
- `internal/cmd/root.go` — wired `newMCPCommand()` into root

## Translation Matrix (6 paths)
| From | To |
|------|----|
| Copilot | Claude, OpenCode |
| Claude | Copilot, OpenCode |
| OpenCode | Copilot, Claude |

## Design
1. **Reader** parses each assistant's JSON into canonical `Config`
2. **Translator** (existing) converts canonical to target format
3. **Merger** (existing) handles add/skip/conflict semantics
4. **Translate()** orchestrates the full pipeline

Closes #79

### CODE REVIEW

File: internal/cmd/mcp_translate.go [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+				return fmt.Errorf("--from is required")
+			}
+			if toRaw == "" {
+				return fmt.Errorf("--to is required")
+			}
+
+			targets := strings.Split(toRaw, ",")
+			for i := range targets {
+				targets[i] = strings.TrimSpace(targets[i])
+			}
```
[the remark this answers is not in the document]
Comment (willvelida) [reply internal/cmd/mcp_translate.go]: Good catch — fixed in 6c7911c. Empty strings from comma-separated \--to\ values are now filtered out, and we return a clear error if no valid targets remain after filtering.
